lib/validator/v3/validator.go

```diff
@@ -0,0 +1,729 @@
+// Copyright (c) 2016-2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	"net"
+	"reflect"
+	"regexp"
+	"strings"
+
+	api "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/numorstring"
+	"github.com/projectcalico/libcalico-go/lib/scope"
+	"github.com/projectcalico/libcalico-go/lib/selector"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"gopkg.in/go-playground/validator.v8"
+)
+
+var validate *validator.Validate
+
+var (
+	labelRegex            = regexp.MustCompile("^[a-zA-Z0-9_./-]{1,512}$")
+	labelValueRegex       = regexp.MustCompile("^[a-zA-Z0-9]?([a-zA-Z0-9_.-]{0,61}[a-zA-Z0-9])?$")
+	nameRegex             = regexp.MustCompile("^[a-zA-Z0-9_.-]{1,128}$")
+	labelsToApplyValRegex = regexp.MustCompile("^[a-zA-Z0-9_.-]{0,128}$")
+	namespacedNameRegex   = regexp.MustCompile(`^[a-zA-Z0-9_./-]{1,128}$`)
+	interfaceRegex        = regexp.MustCompile("^[a-zA-Z0-9_-]{1,15}$")
+	actionRegex           = regexp.MustCompile("^(Allow|Deny|Log|Pass)$")
+	backendActionRegex    = regexp.MustCompile("^(Allow|Deny|Log|next-tier|)$")
+	protocolRegex         = regexp.MustCompile("^(TCP|UDP|ICMP|ICMPv6|SCTP|UDPLite)$")
+	ipipModeRegex         = regexp.MustCompile("^(Always|CrossSubnet|Never)$")
+	bgpLogLevel           = regexp.MustCompile("^(Info|Warning|Debug|Error|Fatal|Trace)$")
```

Comment:
Still not sure where Trace comes from for bgpLogLevel.

nit:  Can we either order alphabetically or in order of severity :-)  (it was probably already like this - in which case I'm prolly the one to blame)

Reply:
Will reorder from most severe to least: `(Fatal|Error|Warning|Info|Debug)`
Also sounds like `Trace` isn't supported/necessary, so I'll combine Felix and Bgp logging regexes. I'll leave the validation tags and functions separate for now, but we can combine them if necessary later.